File: Three/Three.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <utility>
using namespace std;
long long const MAX_N = 100000;
long long MAX_A = 2000000000;
long long N;
long long K;

int main()
{
    long long modul = 1e9 + 7;
    cin >> N;
    cin >> K;
    vector<pair<long, int>> seq(N);
    vector<long> kp;
    vector<long> km;
    bool havepositive = false;
    for (long long i = 0; i < N; ++i)
    {
        cin >> seq[i].first;
        if (seq[i].first > 0)
            seq[i].second = 0;
        else
            seq[i].second = 1;

        if (seq[i].first > 0 && !havepositive)
            havepositive = true;
    }

    for (size_t i = 0; i < seq.size(); ++i)
    {
        if (seq[i].second == 1)
            seq[i].first *= -1;
    }

    long long t = seq.size() - 1;
    sort(seq.begin(), seq.end());
    for (long long i = 0; i < K; ++i)
    {
        if (!seq[t].second)
            kp.push_back(seq[t].first);
        else
            km.push_back(seq[t].first);

        t--;
    }

    if (km.size() % 2 == 1)
    {
        if (kp.empty())
        {
            if (havepositive)
            {
                long long maxp = 0;
                km.pop_back();
                for (long long i = t; i >= 0; --i)
                {
                    if (seq[i].first > maxp && !seq[i].second)
                        maxp = seq[i].first;
                }

                kp.push_back(maxp);
            }
            else
            {
                km.clear();
                for (long long i = 0; i < K; i++)
                    km.push_back(seq[i].first);
                km[0] *= -1;
            }
        }
        else
        {
            long long maxp = 0;
            long long maxm = 0;
            long long m;
            long long p;
            if (!kp.empty())
            {
                p = kp[kp.size() - 1];
                kp.pop_back();
            }
            else
                p = 0;

            if (!km.empty())
            {
                m = km[km.size() - 1];
                km.pop_back();
            }
            else
                m = 0;

            for (long long i = t; i >= 0; i--)
            {
                if (!seq[i].second)
                {
                    if (seq[i].first > maxp)
                        maxp = seq[i].first;
                }
                else
                {
                    if (seq[i].first > maxm)
                        maxm = seq[i].first;
                }
            }

            if (maxm * m >= maxp * p)
            {
                km.push_back(m);
                km.push_back(maxm);
            }
            else
            {
                kp.push_back(p);
                kp.push_back(maxp);
            }
        }
    }

    long long multiply = 1;
    for (long long el : kp)
        multiply *= el;

    for (long long el : km)
        multiply *= el;
    
    if (abs(multiply) < modul)
    {
        multiply %= modul;
        if (multiply < 0)
            multiply += modul;   
    }

    while (abs(multiply) >= modul)
    {
        multiply %= modul;
        if (multiply < 0)
            multiply += modul;
    }

    cout << multiply;
    return 0;
}

File: Threeruby/Three.rb
mod = 1000000000 + 7

input = gets.chomp.split.map(&:to_i)
N = input[0]
K = input[1]
seq = gets.chomp.split.map(&:to_i)
km = Array.new()
kp = Array.new()
havepositive = false

for i in 0...N
  a = seq.delete_at(i).to_i
  b = 0
  if a > 0
    b = 0
  else
    b = 1
  end
  seq.insert(i, [a, b])
  if seq[i][0].to_i > 0 and !havepositive
    havepositive = true
  end
end

for i in 0...N
  if seq[i][1] == 1
    seq[i][0] *= -1
  end
end

seq.sort!
t = seq.length.to_i - 1
for i in 0...K
  if (seq[t][1] == 0)
    kp.push(seq[t][0])
  else
    km.push(seq[t][0])
  end
  t -= 1
end

if (km.length.to_i % 2 == 1)
  if (kp.empty? == true)
    if (havepositive == true)
      maxp = 0
      km.pop
      while t >= 0
        if (seq[i][0] > maxp and seq[i][1] == 0)
          maxp = seq[i][0]
        end
        t -= 1
      end
      kp.push(maxp)
    else
      km.clear
      for i in 0...K
        km.push(seq[i][0])
      end
      km[0] *= -1
    end
  else
    maxp = 0
    maxm = 0
    m = 0
    pos = 0
    if kp.empty? == false
      pos = kp[kp.length.to_i - 1]
      kp.pop
    else
      pos = 0
    end

    if km.empty? == false
      m = km[km.length.to_i - 1]
      km.pop
    else
      m = 0
    end

    while t >= 0
      if seq[t][1] == 0
        if seq[t][0] > maxp
          maxp = seq[t][0]
        end
      elsif seq[t][0] > maxm
        maxm = seq[t][0]
      end
      t -= 1
    end

    if maxm * m >= maxp * pos
      km.push(m)
      km.push(maxm)
    else
      kp.push(pos)
      kp.push(maxp)
    end
  end
end

multiply = 1
for el in kp
  multiply *= el
end
for el in km
  multiply *= el
end

multiply.to_i

# if multiply > 0
#   multiply %= mod
# else
#   multiply *= -1
multiply %= mod
#   multiply *= -1
# end
print(multiply)
# print ("kp - > \n")
# for el in kp
#   print el
#   print ", "
# end
# print "\n km - > \n"

# for el in km
#   print el
#   print ", "
# end
